File: Tyche.iOS/DataPool/Sources/DataPool/Infrastructure/PoolRemoteRepository.swift
import Core

class PoolRemoteRepository : PoolRepository {
    private let poolRemoteDataSource: PoolRemoteDataSource

    private let networkErrorHandler: NetworkErrorHandler

    init(
        poolRemoteDataSource: PoolRemoteDataSource,
        networkErrorHandler: NetworkErrorHandler
    ) {
        self.poolRemoteDataSource = poolRemoteDataSource
        self.networkErrorHandler = networkErrorHandler
    }

    func getPool(id: String) async -> Result<Pool, any Error> {
        return await networkErrorHandler.handle {
            let poolResponse = try await poolRemoteDataSource.getPool(id: id)
            return poolResponse.toPool()
        }
    }

    func createPool(createPoolInput: CreatePoolInput) async -> Result<CreatePoolOutput, any Error> {
        return await networkErrorHandler.handle {
            let createPoolResponse = try await poolRemoteDataSource.createPool(creaePoolRequest: createPoolInput.toCreatePoolRequest())
            return createPoolResponse.toCreatePoolOutput()
        }
    }

    func joinPool(joinPoolInput: JoinPoolInput) async -> Result<Void, any Error> {
        return await networkErrorHandler.handle {
            try await poolRemoteDataSource.joinPool(joinPoolRequest: joinPoolInput.toJoinPoolRequest())
        }
    }
}

File: Android/data/pool/src/main/java/com/felipearpa/tyche/data/pool/infrastructure/PoolRemoteRepository.kt
package com.felipearpa.tyche.data.pool.infrastructure

import com.felipearpa.network.NetworkExceptionHandler
import com.felipearpa.tyche.data.pool.domain.CreatePoolInput
import com.felipearpa.tyche.data.pool.domain.CreatePoolOutput
import com.felipearpa.tyche.data.pool.domain.JoinPoolInput
import com.felipearpa.tyche.data.pool.domain.Pool
import com.felipearpa.tyche.data.pool.domain.PoolDataSource
import com.felipearpa.tyche.data.pool.domain.PoolRepository
import com.felipearpa.tyche.data.pool.domain.toCreatePoolOutput
import com.felipearpa.tyche.data.pool.domain.toCreatePoolRequest
import com.felipearpa.tyche.data.pool.domain.toJoinPoolRequest
import com.felipearpa.tyche.data.pool.domain.toPool

internal class PoolRemoteRepository(
    private val poolDataSource: PoolDataSource,
    private val networkExceptionHandler: NetworkExceptionHandler,
) : PoolRepository {
    override suspend fun getPool(id: String): Result<Pool> {
        return networkExceptionHandler.handle {
            poolDataSource.getPool(id = id).toPool()
        }
    }

    override suspend fun createPool(createPoolInput: CreatePoolInput): Result<CreatePoolOutput> {
        return networkExceptionHandler.handle {
            poolDataSource.createPool(createPoolRequest = createPoolInput.toCreatePoolRequest())
                .toCreatePoolOutput()
        }
    }

    override suspend fun joinPool(joinPoolInput: JoinPoolInput): Result<Unit> {
        return networkExceptionHandler.handle {
            poolDataSource.joinPool(
                poolId = joinPoolInput.poolId,
                joinPoolRequest = joinPoolInput.toJoinPoolRequest(),
            )
        }
    }
}
